Tools/TestRunnerShared/cocoa/LayoutTestSpellChecker.mm

```diff
@@ -26,107 +26,107 @@
 #import "config.h"
 #import "LayoutTestSpellChecker.h"
 
+#import "InstanceMethodSwizzler.h"
 #import "JSBasics.h"
 #import <objc/runtime.h>
 #import <wtf/Assertions.h>
 #import <wtf/BlockPtr.h>
 
-#if PLATFORM(MAC)
+#if PLATFORM(IOS_FAMILY)
+#import "UIKitSPI.h"
+#endif
 
+#if PLATFORM(MAC)
 using TextCheckingCompletionHandler = void(^)(NSInteger, NSArray<NSTextCheckingResult *> *, NSOrthography *, NSInteger);
+#else
+static NSString *NSGrammarRange = @"NSGrammarRange";
+static NSString *NSGrammarCorrections = @"NSGrammarCorrections";
+#endif
 
-static RetainPtr<LayoutTestSpellChecker> globalSpellChecker;
+static LayoutTestSpellChecker *globalSpellChecker = nil;
 static BOOL hasSwizzledLayoutTestSpellChecker = NO;
+
+#if PLATFORM(MAC)
 static IMP globallySwizzledSharedSpellCheckerImplementation;
 static Method originalSharedSpellCheckerMethod;
+#else
+static IMP globallySwizzledInitializeTextCheckerImplementation;
+static Method originalInitializeTextCheckerMethod;
+
+@protocol TextComposerPostEditorProtocol <NSObject>
+- (NSDictionary<NSString *, id> *)grammarDetailForString:(NSString *)stringToCheck range:(NSRange)range language:(NSString *)language;
+@end
+
+@interface UITextChecker (TestingSupport)
+@property (readonly, nonatomic) id<TextComposerPostEditorProtocol> postEditor;
+@end
+#endif
 
 static LayoutTestSpellChecker *ensureGlobalLayoutTestSpellChecker()
 {
     static dispatch_once_t onceToken;
     dispatch_once(&onceToken, ^{
-        globalSpellChecker = adoptNS([[LayoutTestSpellChecker alloc] init]);
+        globalSpellChecker = [[LayoutTestSpellChecker alloc] init];
     });
-    return globalSpellChecker.get();
-}
-
-static const char *stringForCorrectionResponse(NSCorrectionResponse correctionResponse)
-{
-    switch (correctionResponse) {
-    case NSCorrectionResponseNone:
-        return "none";
-    case NSCorrectionResponseAccepted:
-        return "accepted";
-    case NSCorrectionResponseRejected:
-        return "rejected";
-    case NSCorrectionResponseIgnored:
-        return "ignored";
-    case NSCorrectionResponseEdited:
-        return "edited";
-    case NSCorrectionResponseReverted:
-        return "reverted";
-    }
-    return "invalid";
+    return globalSpellChecker;
 }
 
-static NSTextCheckingType nsTextCheckingType(JSStringRef jsType)
+static NSTextCheckingType nsTextCheckingType(NSString *typeString)
 {
-    auto cfType = adoptCF(JSStringCopyCFString(kCFAllocatorDefault, jsType));
-    if (CFStringCompare(cfType.get(), CFSTR("orthography"), kCFCompareCaseInsensitive) == kCFCompareEqualTo)
+    if ([typeString isEqualToString:@"orthography"])
```

Comment:
Is it OK that these are case sensitive now?

Reply:
I think I'll be okay, since all the existing tests that rely on this all use lowercase keys (I think my original decision to make the comparison case-insensitive was just an arbitrary one).

Thanks for the review!